content/exchange/artifacts/Windows.Sys.LoggedInUsers.yaml

```diff
@@ -25,9 +25,29 @@ sources:
       SELECT OS From info() where OS = 'windows'
 
     query: |
-      LET CurrentlyLoggedIn = SELECT parse_string_with_regex(string=Antecedent,regex=['Domain=\\"(.*?)\\"']).g1 as Domain,parse_string_with_regex(string=Antecedent,regex=['Name=\\"(.*)\\"']).g1 as LogonName, parse_string_with_regex(string=Dependent,regex=['LogonId=\\"([0-9]+)\\"']).g1 as CurrentLogonId FROM wmi(query="SELECT * FROM win32_loggedonuser", namespace="ROOT/CIMV2") WHERE LogonName =~ UserNameRegex
+    
+        // Helper functions
+        LET TimeFormat = "20060102150405.999999-0700"
 
-      LET Sessions = SELECT LogonName,Domain,LogonId,LogonType,timestamp(string=parse_string_with_regex(string=StartTime,regex=['([0-9\\.]+).*']).g1+format(args=atoi(string=parse_string_with_regex(string=StartTime,regex=['(-|\\+)([0-9]+)']).g2)/60,format=parse_string_with_regex(string=StartTime,regex=['.*(\\+|-)']).g1+"%02.0f00"),format="20060102150405.999999-0700") as StartTimestamp
-      FROM wmi(query="SELECT * FROM Win32_LogonSession", namespace="ROOT/CIMV2") WHERE LogonId=CurrentLogonId AND Domain =~ DomainRegex AND LogonType =~ LogonTypeRegex
-
-      SELECT * FROM foreach(row=CurrentlyLoggedIn, query=Sessions)
+        LET ExtractDomain(Antecedent) = parse_string_with_regex(string=Antecedent,regex=['Domain=\\"(.*?)\\"']).g1
+        
+        LET ExtractLogonName(Antecedent) = parse_string_with_regex(string=Antecedent,regex=['Name=\\"(.*)\\"']).g1
+        
+        LET ExtractLogonID(data) = parse_string_with_regex(string=data,regex=['LogonId=\\"([0-9]+)\\"']).g1
+        
+        LET FormatTime(StartTime) = timestamp(string=parse_string_with_regex(string=StartTime,regex=['([0-9\\.]+).*']).g1+format(args=atoi(string=parse_string_with_regex(string=StartTime,regex=['(-|\\+)([0-9]+)']).g2)/60,format=parse_string_with_regex(string=StartTime,regex=['.*(\\+|-)']).g1+"%02.0f00"),format=TimeFormat) 
```

Comment:
This should not be necessary. The timestamp function accepts a format string to control parsing

Reply:
I think we can get pretty close using this:
```
LET TimeFormat = "20060102150405.999999-0700"
timestamp(string=regex_replace(source=Time,replace="-0",re="-"),format=TimeFormat)
```

But this format `20241026201539.494314-420` doesn't seem to be easy to parse with Go's time.Parse, specifically the -420 minutes offset.

Reply:
You are right - this non standard time format is really a pain to work with. I came up with 

```sql
LET _X(X) = parse_string_with_regex(regex="(^.+)(-\\d+)$", string=X)
LET NormalizeTime(X) = format(format="%s%03g00", args=[_X(X=X).g1, int(int=_X(X=X).g2) / 60])
LET ParseTime(X) = timestamp(string=NormalizeTime(X=X), format="20060102150405.999999-0700")

SELECT ParseTime(X="20241026201539.494314-420")
FROM scope()
```

Do you think thats more readable than what you had initially?

Reply:
That's much better, thanks! Just pushed a commit using this method and with indents.